### PR TITLE
Validate Host DNS name in BareMetalHost resource

### PR DESCRIPTION
**Pull request by hroyrh**
Validate host dns name part of BareMetalHost bmc address.

### CODE REVIEW

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 1]
```diff
@@ -53,3 +60,24 @@ func validateRAID(r *RAIDConfig) error {
 
 	return nil
 }
+
+func validateDNSName(hostaddress string) error {
+
+	if hostaddress == "" {
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:66]: Is this valid?

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 2]
```diff
@@ -53,3 +60,24 @@ func validateRAID(r *RAIDConfig) error {
...
+		return nil
+	}
+
+	hurl, _ := url.Parse(hostaddress)
+	host, _, err := net.SplitHostPort(hurl.Host)
+
+	if err != nil {
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:73]: Maybe error is not nil actually means that host is invalid?

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 3]
```diff
@@ -53,3 +60,24 @@ func validateRAID(r *RAIDConfig) error {
...
+		host = hurl.Host
+	}
+	valid, _ := regexp.MatchString(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`, host)
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:76]: I think, we should also handle error case in here instead hiding it.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 4]
```diff
@@ -53,3 +60,24 @@ func validateRAID(r *RAIDConfig) error {
...
+
+	if !(valid) {
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:78]: There is no need for parenthesis in here.

File: apis/metal3.io/v1alpha1/baremetalhost_validation_test.go [hunk 5]
```diff
@@ -71,6 +71,17 @@ func TestValidateCreate(t *testing.T) {
 			oldBMH:    nil,
 			wantedErr: "hardwareRAIDVolumes and softwareRAIDVolumes can not be set at the same time",
 		},
+		{
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation_test.go:74]: I think, we can add more test like happy path or more edge cases.

**Review by ardaguclu — COMMENTED** [review]
Thanks @hroyrh for the PR

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 1 above)
Comment (hroyrh) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:66]: This is to validate empty BareMetalHostSpec, but there could be a more precise check - maybe check if address is specified when Spec is not empty ?

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 2 above)
Comment (hroyrh) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:73]: Yes, I should probably check if the port is empty rather than the error.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 1 above)
Comment (ardaguclu) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:66]: I mean; if hostaddress is empty, should we return error or return nil

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 2 above)
Comment (ardaguclu) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:73]: İf error is not nil, we are setting `host = hurl.host`, why we are doing that instead returning an error?.

File: apis/metal3.io/v1alpha1/baremetalhost_validation_test.go (hunk 5 above)
Comment (hroyrh) [reply apis/metal3.io/v1alpha1/baremetalhost_validation_test.go:74]: Okay, I will add a few more test cases.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 1 above)
Comment (zaneb) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:66]: I can potentially see some value in creating a Host with no address and then specifying the address later. I can't think of any particular use cases though... the BMO can't do anything until there is an address.
tl;dr I don't know :laughing:

Comment (ardaguclu) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:66]: It might be safe to return nil then :)

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 6]
```diff
@@ -53,3 +60,24 @@ func validateRAID(r *RAIDConfig) error {
...
 }
+
+func validateDNSName(hostaddress string) error {
+
+	if hostaddress == "" {
+		return nil
+	}
+
+	hurl, _ := url.Parse(hostaddress)
+	host, port, _ := net.SplitHostPort(hurl.Host)
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:71]: @hroyrh, if port is empty `host = hurl.Host` seems ok to me. But why we are not checking error in here?

Comment (hroyrh) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:71]: It raises error for port empty too and as for the validity of the url, the regex checks that.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 7]
```diff
@@ -18,6 +21,10 @@ func (host *BareMetalHost) validateHost() []error {
 		errs = append(errs, err)
 	}
 
+	if err := validateDNSName(host.Spec.BMC.Address); err != nil {
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:24]: @hroyrh hardwareutils/bmc package is added in this module. Maybe it would be useful using bmc's functionality for that validation if it is possible.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 8]
```diff
@@ -53,3 +60,24 @@ func validateRAID(r *RAIDConfig) error {
 
 	return nil
 }
+
+func validateDNSName(hostaddress string) error {
```
Comment (ardaguclu) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:64]: @hroyrh what I was thinking is something like that;

```
if hostaddress == "" {
		return nil
	}

	hurl, err := url.Parse(hostaddress)
	if err != nil {
		if !strings.Contains(hostaddress, ":") {
			return fmt.Errorf("Invalid format")
		}
	}
	host, port, err := net.SplitHostPort(hurl.Host)
	if err != nil {
		return fmt.Errorf("another invalid format")
	}

	if port == "" {
		host = hurl.Host
	}
	valid, err := regexp.MatchString(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`, host)
	if err != nil {
		return fmt.Errorf("regex error")
	}

	if !valid {
		return fmt.Errorf("Host DNS name is invalid")
	}

	return nil
```

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 7 above)
Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:24]: I agree, the validation itself should probably go to the bmc package, if it's not there already

**Comment by kashifest** [comment]
/hold
Holding for a while. Changing the default branch reference to `main` from `master`. Once the transition is done I will unhold it.

**Comment by kashifest** [comment]
/hold cancel
The switch is done. BMO's default branch is now `main`

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 9]
```diff
@@ -143,3 +147,17 @@ func validateBMHName(bmhname string) error {
...
+
+	if hostaddress == "" {
+		return nil
+	}
+
+	_, err := bmc.GetParsedURL(hostaddress)
+	if err != nil {
+		return fmt.Errorf("host DNS name is invalid")
```
Comment (dtantsur) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:159]: Could you include the original error using `errors.Wrap`?

Comment (hroyrh) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go]: @dtantsur I've wrapped the errors, and added a few more test cases, can you please take another look.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 10]
```diff
@@ -143,3 +148,17 @@ func validateBMHName(bmhname string) error {
...
+
+func validateDNSName(hostaddress string) error {
+
+	if hostaddress == "" {
+		return nil
+	}
+
+	_, err := bmc.GetParsedURL(hostaddress)
+	if err != nil {
+		return errors.Wrap(err, "BMO validation")
```
[the remark this answers is not in the document]
Comment (hroyrh) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:160]: Actually, the wrapped error already contains something very similar plus some other details, example error message : `BMO validation: failed to parse BMC address information: BMC address hostname/IP : [-host.example.com.org] is invalid` ( see [line 394](https://github.com/metal3-io/baremetal-operator/pull/1040/files#diff-c63bbb835e932967f86894fa9e1ca2d32aa663f19c2ce0712f53e0b65ebd3bf9R394) in the test file )

[the remark this answers is not in the document]
Comment (dtantsur) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:160]: Nice!
